### ISSUE TITLE
Func extensions install

### ISSUE THREAD

**Issue by asavaritayal** [Microsoft/vscode-azurefunctions#382]
Functions now has capabilities that let the user run `func extension install` at the base of a function project. This will inspect the function.json files and install the relevant extensions. Would be a good to have an experience for the same in VS Code. Perhaps, an option in the **Command Palette** ? @fiveisprime 

Note: This applies to all languages other than C# class library (.net)

**Comment by asavaritayal** [comment]
@jeffhollan @soninaren can you confirm which version of core tools this works with?

**Comment by jeffhollan** [comment]
@fiveisprime from my usage the command seems pretty low cost, so I expect adding it into the launch.json config for the project to run before any debugging may be pretty cheap to do.  Should also likely look at doing as a pre-req to publish as well (in case someone wrote a valid function and wanted to publish without ever debugging).

Also if it matters Java doesn't have to have this support as the maven plugin that powers Java dev does this as part of the `mvn package` step.  So Java and C# don't need this added.

**Comment by EricJizbaMSFT** [comment]
@jeffhollan my only concern with adding this to every project is that the extensions are based on .NET Core, correct? Does running `func extension install` require users to have the dotnet cli installed?

Even if it doesn't break our users, I'm also a little concerned that it will confuse some of our node users that haven't been using extensions and then all of the sudden they have a 'csproj' file in their project. @fiveisprime thoughts on this? Are you okay with the possible confusion?

**Comment by jeffhollan** [comment]
@EricJizbaMSFT this is correct - the command could be run but for it to work it does require .NET Core SDK installed on machine.  

That said there is a new work item we are exploring led by @paulbatum and @soninaren to use the core tools alone to pull down the extensions (or include the latest in the runtime).  It's still being explored so can't commit - but motivation is to address a lot of concerns you have.  

Hopefully that comes about but if not we definitely need some better tooling around extension installation in next month or two. I'm assuming work to add `func extension install` is relatively small? I'd prefer we either add now and potentially pull, or wait a bit more to see if this new work item may pan out -- but if not still provide enough time to do the current suggestion (assuming @fiveisprime is ok) so we at least have something better than now.

**Comment by EricJizbaMSFT** [comment]
@jeffhollan any update on this? Is there an issue tracking the work to run `func extensions install` without .NET Core installed on the machine?

**Comment by jeffhollan** [comment]
@soninaren was looking into this. Last I heard (earlier this week) was that investigation was still underway into feasibility and complexity and not quite sure what form will take.  I can let him link to any potential update or discussion on GitHub around this.

I do still get frequent pings (one earlier this week) from JavaScript users though not understanding some of our triggers are actually available in v2 runtime (like Event Hubs trigger) which largely stems from them needing to run this extension install step.  We are likely getting close to the time where we may need to assume things will stay as-is (not optimal) but help make VS Code tooling do things like auto-run extensions install as part of debug or publish, and potentially prompt to download .NET Sdk if not installed (or point to a doc we can write explaining why).  Goal is that experience would be better than today by GA regardless if best (no dependency on .NET SDK) or better (guide user through current extension requirements)

**Comment by jeffhollan** [comment]
HTTP is the most popular for active or inactive. For active the most popular are queues and event hubs.

The reason it's not needed for C# projects is that what `func extensions install` is doing is creating a project and defining the dependencies so the right binaries get pulled into the project.  With C# you already have a project file that already has those dependencies listed in it for it to work at all. Not to mention func extensions install works by looking at the `function.json` file which only Python + JavaScript projects will have users manually editing.

In short it's unfortunate we have to do this - I'm not saying it's the best case, but it's the reality we are faced and if we don't try to make this easier from a tooling standpoint a lot of people will be confused as to why their stuff doesn't work.

**Comment by EricJizbaMSFT** [comment]
Ah makes sense for C#.

In any case - is is possible to have the "no-op" for http-only projects?

**Comment by soninaren** [comment]
It is already a no-op for http-only projects. http trigger doesn't require extension installation. So for http-only project cli will end finding nothing to install. i can double check this and optimize this experience.

**Comment by jeffhollan** [comment]
I think that makes sense - but note that an HTTP function that uses like a CosmosDB output binding would still need the extension.  That said I expect those smarts would need to be in the tooling - not sure what the cli would know about it

**Comment by jeffhollan** [comment]
Yep makes sense.  Spoke to @soninaren and he mentioned he can make update in next CLI so if only bindings being used are HTTP or Timer it won't touch or change anything on the project

**Comment by EricJizbaMSFT** [comment]
Perfect thanks. Is there an issue to track that work?

Just to confirm - this is required before deploying as well?

**Comment by jeffhollan** [comment]
Yes likely safest to run before deploying.  @soninaren can you create an issue on the core tools repo and link?

**Comment by EricJizbaMSFT** [comment]
@jeffhollan what about local git deploy? Do users need to check the "extensions.csproj" files in to git? and will a `dotnet restore` be handled for them?

**Comment by jeffhollan** [comment]
Yes extensions.csproj should be checked into git as it would ensure other developers are working with the same version of the extension.  running `func extensions install` and `func extensions sync` after GA will honor existing project and restore the version they initially created with.  there would be another gesture to update extensions once / when that occurs

**Comment by jeffhollan** [comment]
The bin/ folder that is generated SHOULD NOT be checked into source control - would be same as package.json vs node_modules folders

**Comment by EricJizbaMSFT** [comment]
Okay thanks. As as side point, though - I wasn't aware of `func extensions sync` - should we be running that for the user as well? Before debug and/or deploy? Not sure about the difference between 'install' and 'sync'

**Comment by jeffhollan** [comment]
Today in preview, `install` will add missing extensions, upgrade existing, or install defined extensions.  `sync` today will only install what is defined (won't add missing or upgrade).  **however** when we GA we will remove the auto-upgrade behavior of install.  So I think for you, just using install is best and very possible we may deprecate `sync` as only difference could be it won't add missing extensions (which is likely what you want to happen anyway)

**Comment by soninaren** [comment]
Created an issue on core-tools to make sure `func extensions install` is a no-op 
https://github.com/Azure/azure-functions-core-tools/issues/688
